### ISSUE TITLE
[RFC] 1 WIP: Docs as Code

### ISSUE THREAD

**Issue by frankhinek** [TBD54566975/tbd-tutorials#4]
- RFC Name: Docs as Code
 - RFC ID: 1
 - Start Date: 2022-08-22
 - Contributors: @frankhinek, @ALRubinger, @saeedjabbar
 - Current Status: WIP

# Summary

Proposal to adopt a docs-as-code approach for developing a technical documentation system that helps orient newcomers, helps contributors acquire and apply knowledge, and helps create an inclusive community.

# Context and Scope

TBD's projects are developed in the public guided by an [open source philosophy](https://developer.tbd.website/open-source) with three foundational pillars: inclusivity, transparency, and clarity.  This proposal is focused on making progress towards fulfilling these commitments:

> - Anyone with the capacity to contribute: can.
> - We are proactive in welcoming a diverse contributor base for our code, documentation, ...
> - Decision-making is open by default, ...
> - Projects need ... accessible, comprehensive documentation, a stellar Getting Started experience, and well-defined expectations.

"Incomplete or confusing documentation" has [frequently been](https://opensourcesurvey.org/2017/) [cited](https://octoverse.github.com/creating-documentation/) as one of the top problems in open-source software development.  Accurate, up-to-date, and well-organized documentation improves the first time contributor experience, creates a more inclusive environment, and can result in [significant productivity boosts](https://octoverse.github.com/creating-documentation/#docs-for-productivity-and-culture).

Given the target audience for [developer.tbd.website](https://developer.tbd.website) and early stage of development for most TBD projects, it follows that the documentation should be optimized for the new contributor experience.  Using a docs-as-code approach combines the power of developer tools with techniques to encourage community engagement and improve documentation quality.

Treating docs-as-code means that document publishing follows a similar process to software development:
- documentation is stored as plain text in a version control system
- issues and pull requests track all documentation changes and decisions
- additions and revisions are reviewed
- approved documents are published without much human intervention

By taking an open approach that is accessible and familiar to developers, TBD can establish a virtuous cycle by encouraging more community enhancements and feedback to improve the contributor experience, which in turn leads to more engagement.

## Goals

- Anyone in the community can contribute enhancements, fixes, and ideas.
- Documentation source files are stored in a publicly accessible repository.
- Documentation and code/command examples in separate files.
- Minimize the dependencies and onboarding steps for contributors.
- The documentation repository is organized in a thoughtfully structured manner.
- Changes to documents are reviewed and must be approved prior to publication.
- All code and command examples are manually tested during review.
- Once approved, document processing and artifact builds are automated.
- Documentation is published to [developer.tbd.website](https://developer.tbd.website).

## Non-goals

Several of these could reasonably be goals, but are explicitly chosen not to be goals in favor of releasing a minimum viable documentation system and receiving feedback from the community.

- Definition of documentation templates
- Automated code/command testing and linting
- Multi-language support
- Search
- Support for JSX in Markdown documents (.mdx)

# Proposal

## Document Format

- Documentation text is stored as plain text files.
- Markdown formatting is used to add style to plain text.
- [CommonMark](https://commonmark.org) syntax specification for Markdown.
- Code/command examples are stored separately from documentation text to minimize migration friction if/when automated testing is introduced.

## Documentation Processing

- Target output format for inclusion in [developer.tbd.website](https://developer.tbd.website) is Markdown so that the Docusaurus-powered site can import post-processed Markdown files.
- Use [remark](https://github.com/remarkjs/remark) to parse, transform, and serialize Markdown content. [Why remark?](#why-remark)
- A new command-line utility, [tentatively tbdoc](https://github.com/frankhinek/prototype-tbdocs/tree/main/tbdoc), automatically transforms documentation source files into artifacts ready to be imported to [developer.tbd.website](https://developer.tbd.website).  [Why wrap with Docker?](#why-docker)

## Repository Structure

- Developed and maintained by the community on GitHub in a single @TBD54566975 repository.
- Documentation is organized by topic to improve discovery, authoring, and maintenance (e.g., SSI Service, DWN, third-party integrations, etc.).
- Each topic consists of:
    - Sub-directories for tutorials and how-to guides.
    - If appropriate, one or more Dockerfiles which are shared by topic documentation.  Ideally, the Docker image definition maintained within the associated TBD project repo can be used to minimize maintenance.
    - A sub-directory is created for each tutorial or how-to guide and consists of:
        - Markdown files that contain the documentation text.
        - Images that support the content.
        - A `workflows` directory that contains an individual file for each command or code example.  During the documentation build process, these files will be parsed and inserted inline with the documentation text.

    ![Repository Structure](https://user-images.githubusercontent.com/134693/189328225-845e940f-b34c-493a-8d68-54772ebaf478.svg)

    "[Overview of the Galaxy Training Material](https://training.galaxyproject.org/training-material/topics/contributing/slides/introduction.html)" by [Bérénice Batut](https://training.galaxyproject.org/training-material/hall-of-fame/bebatut/), used under [CC BY](https://creativecommons.org/licenses/by/4.0/) / Redrawn and colorized

## Documentation Flows

Document publishing follows a similar process to software development:

```mermaid
%%{init: { 'logLevel': 'debug', 'theme': 'base', 'gitGraph': {'showBranches': true, 'showCommitLabel':false,'mainBranchName': 'main'}} }%%
      gitGraph
        commit
        commit tag:"Create Branch"
        branch new-topic
        commit
        commit
        commit type: HIGHLIGHT tag:"Open PR"
        commit
        commit
        commit tag:"Review & Improvement"
        commit
        checkout main
        merge new-topic tag:"PR Merged"
        commit
        commit
```

### Authoring New Content

1. Contributor forks the TBD community docs repo on GitHub and creates a new branch.
3. Contributor creates the directory structure scaffolding for a new tutorial.
4. Contributor authors the document, including any command/code and/or images.
5. Contributor opens a pull request (PR) in the TBD community docs repo.
6. TBD and community members review, manually test command/code examples, and provide feedback.
7. If approved, the PR is merged to the TBD community docs repo.
8. GitHub Actions automatically process the contributed docs and open a PR on the TBD Developer Site.

### Suggesting a Fix or Improvement

1. Contributor opens an Issue or PR in the TBD GitHub repo.
2. TBD and community members review and provide feedback.
4. If approved, a PR is created (if starting as an Issue) and approved/merged.
5. Automated workflow publishes the revised content.

# Definition of success

- Community contributors author, submit, and have their content published to [developer.tbd.website](developer.tbd.website).
- Positive feedback is received from the community.
- The docs-as-code approach and design proposal can be adapted to automate publishing to [developer.tbd.website](developer.tbd.website).
- This design can be enhanced and extended over time as a function of being modular and flexible.

# Open Questions

1. After a community contribution is approved and merged into the community docs repo, a GitHub Action will automatically create a PR in the TBD Developer Site.  When this action is creating the PR, should it create a new branch and PR for each approved push to the community docs repo?  Or should any additional pushes to the community docs repo main branch be added to the open TBD Developer Site PR until it is approved or closed?

# Appendix

## Alternatives Considered

- [MarkdownPP](https://github.com/jreese/markdown-pp)
- [Pandoc 3](https://pandoc.org)
- [Antora](https://antora.org)
- [Gitdown](https://github.com/gajus/gitdown)
- [Markdoc](https://markdoc.io)
- [Postman API Platform](https://www.postman.com/api-platform/)
- [Gitbook](https://www.gitbook.com)
- [Mkdocs](https://www.mkdocs.org)
- [ReadMe](https://readme.com)

## Design Decision Rationale

<h3 dir="auto" id="why-remark">Why Remark?</h3>

- Can be easily extended with linting, grammar, and language use plugins to drive up the quality of the documentation and make it more accessible to groups that are underrepresented in open source software development.
- Markdown parsing engine used by Docusaurus, MDX, Node.js docs, GitHub Docs, and [many other projects](http://unifiedjs.com/community/case/).
- The [most downloaded](https://www.npmtrends.com/remark-parse-vs-marked-vs-micromark-vs-markdown-it) Markdown parser.
-  100% [compliant](https://github.com/remarkjs/remark#syntax) to CommonMark and 100% to GitHub Flavored Markdown (GFM) with a plugin.
- Hundreds of available [plugins](https://github.com/remarkjs/remark#plugins) to support a broad variety of use cases and workflows.

<h3 dir="auto" id="why-docker">Why encapsulate utility tools/scripts with Docker?</h3>

- One of the high priority goals for this proposal is that _Anyone in the community can contribute..._, and to that end, a secondary objective is to _Minimize the dependencies and onboarding steps for contributors_.  By encapsulating utility tools/scripts using a Docker container, the only dependency a contributor must satisfy is a local installation of a Docker engine.
- For documentation contributors whose strengths are in technical writing or who aren't polyglot programmers, the major benefit to this approach is eliminating onboarding steps such as installing runtime environments (e.g., Node.js, JVM, etc.), modifying the `PATH` environment variable, or managing multiple runtime environment versions needed to satisfy different project requirements (e.g., Node 10 vs. 16).
- Docker containers run applications in a consistent, isolated environment which helps to ensure that more contributors are successful when using the tools and less community time is spent troubleshooting.

## Decision Log

1. Docusaurus is used for [developer.tbd.website](https://developer.tbd.website), which supports [MDX](https://mdxjs.com/docs/what-is-mdx/).  Should we support including JSX in markdown content for community-contributed docs?

    **DECISION**: MDX/JSX will not be supported in the initial release.

    **Additional Context**: There may be complications related to how Docusaurus handles MDX and incorporates external Markdown. Long term we may want to offer docs authors greater flexibility in styling by exposing JSX directly. In service of MVP, we will limit the scope to exclude support for community contributors authoring MDX and release a system which lets us author in Markdown, embed code and images and publish content on the TBD Developer Site.

2. When new documentation is approved and PRs merged, GitHub Actions can be triggered to build the doc artifacts and potentially trigger publishing to [developer.tbd.website](https://develop.tbd.website).  Should this workflow run in the main Docusaurus site repo or the community-contributed docs repo?  What additional steps can we take to automate the publication process and minimize human intervention?

    **DECISION**: The community-contributed docs repo will use GitHub Actions to build the output Markdown and trigger a Pull Request on the TBD Developer Site repo.

    **Additional Context**: The community-contributed docs repo can run builds to compose the output Markdown, and then we trigger a PR into the Developer Site repo to update/add the newly-generated page? This would let us always revert the Developer Site if we needed, and makes clear when a new piece of content is to be published, while letting the docs repo handle its own build to generate the markdown.

**Comment by ALRubinger** [comment]
This is a remarkable effort. Thank you, @frankhinek!

Overall, it's a path I'd love to continue exploring. I appreciate the Docker-based approach to lock the build environment and give devs consistency. Remark looks like it's working great so far for the code injection and templating. My run through the docs here got me a clean build with no issues and the `sh` snippits inlined into the output. Overall, I'm really curious to see how this works if we do automate this into a PR for the developer site - I think that'll be proof that we're on to something. As you call out, the workflow is super important to contribution and to our validating that we've really lowered the bar to docs submissions from folks. TL;DR from me is I think this makes a lot of sense to make progress, and I suspect we'll have plenty of room to fine-tune along the way.

Some more detailed feedback on items in the WIP:

## Non-Goals
Is the intent here simply to keep scope small to deliver an MVP?

* `Automated code/command testing and linting`

Automated testing, to me, is the piece that protects user experience by making sure that the docs as published will work against a specified version of the software. Are we leaving space for this to be solved later, or do you have opinions about the testing of docs that leave this out of scope longer-term?

* `Multi-language support`

One of the best ways I've found to welcome new contributors is to give them an easy way to translate docs into their preferred language, in turn addressing a more global audience and welcoming them. Same question as above: we'll leave space for this to add easy internationalization so that docs can be prepared in N outputs, pulling in the same code snippits that you've externalized in this example?

## Open Questions

* `Docusaurus is used for [developer.tbd.website](https://developer.tbd.website/), which supports [MDX](https://mdxjs.com/docs/what-is-mdx/). Should we support including JSX in markdown content for community-contributed docs?`

This one may get murky as I believe it's tied into how Docusaurus handles MDX and pulling in external Markdown. Long term we may want to offer docs authors greater flexibility in styling by exposing JSX directly. In service of MVP I'd be thrilled to start by releasing a system which lets us author in Markdown, embed code and images as you've shown, and get things published on the Developer Site. WDYT?

* `When new documentation is approved and PRs merged, GitHub Actions can be triggered to build the doc artifacts and potentially trigger publishing to [developer.tbd.website](https://develop.tbd.website/). Should this workflow run in the main Docusaurus site repo or the community-contributed docs repo? What additional steps can we take to automate the publication process and minimize human intervention?`

Hot takes: chained builds? The community-contributed docs repo can run builds to compose the output Markdown, and then we trigger a PR into the Developer Site repo to update/add the newly-generated page? This would let us always revert the Developer Site if we needed, and makes clear when a new piece of content is to be published, while letting the docs repo handle its own build to generate the markdown.

Will also comment on the [forum post](https://forums.tbd.website/t/tutorials-and-examples-system/89/23) about this to address how we wanna keep the momentum up here. @frankhinek - thank you again!!

**Comment by saeedjabbar** [comment]
@frankhinek Excellent job. See my write up here: https://forums.tbd.website/t/tutorials-and-examples-system/89/25?u=saeedjabbar

**Comment by frankhinek** [comment]
> Overall, I'm really curious to see how this works if we do automate this into a PR for the developer site - I think that'll be proof that we're on to something. As you call out, the workflow is super important to contribution and to our validating that we've really lowered the bar to docs submissions from folks.

Agreed -- wiring up the workflow end-to-end will help to validate the approach, and we can work out any minor issues along the way.  To accelerate the implementation and testing we could create a second repo to simulate the developer.tbd.website Docusaurus integration (GitHub Actions & PRs).  Once the approach is validated and the contributor experience is satisfactory, the workflow can be incorporated with the repo tied to the production deployment.

---

> Is the intent here simply to keep scope small to deliver an MVP?
>
>    `Automated code/command testing and linting`

Yes, precisely.  The intent was in the spirit of getting an MVP going per the [forum discussion thread](https://forums.tbd.website/t/tutorials-and-examples-system/89).  As I started writing out the detailed design proposal, it became clear that getting the contributor experience right with automated tests was going to take some iteration.  There are a few ways to approach both in-lining the statements and authoring tests, which can be decoupled from the initial MVP deployment.

I do suggest that we tackle this in the near-term as it does have numerous benefits for the contributor experience.  "Broken" step-by-step documentation is all too common, particularly for fast moving projects.  Additionally, the longer we wait to introduce this feature, the more clean-up work will be needed to address previously published content.

---

> `Multi-language support`
> Same question as above: we'll leave space for this to add easy internationalization so that docs can be prepared in N outputs, pulling in the same code snippits that you've externalized in this example?

That is a great point and could be a good way to encourage more early contributors.  Based on a review of how Docusaurus handles i18n, it appears that this would be fairly easy to plan for in the initial design and incorporate in the short-to-medium term depending on the prioritization by the community and TBD.

### Q: Does the current developer.tbd.website Docusaurus instance already have i18n configured, and if so, can you share a few details that can guide how to structure community contributed docs?

e.g., translation file structure, single-/multi-domain strategy, etc.

---

> **Should we support including JSX in markdown content for community-contributed docs?**
>
> In service of MVP I'd be thrilled to start by releasing a system which lets us author in Markdown, embed code and images as you've shown, and get things published on the Developer Site. WDYT?

Agreed -- progressing to an initial working MVP seems appropriate.

---

> **Should this workflow run in the main Docusaurus site repo or the community-contributed docs repo?**
> 
> Hot takes: chained builds? The community-contributed docs repo can run builds to compose the output Markdown, and then we trigger a PR into the Developer Site repo to update/add the newly-generated page? This would let us always revert the Developer Site if we needed, and makes clear when a new piece of content is to be published, while letting the docs repo handle its own build to generate the markdown.

I like it for the reasons you highlighted, and it also allows the two to run in parallel at a different pace, if needed.

**Comment by frankhinek** [comment]
## Updates:
* Closed two Open Questions and updated the **Decision Log** based on the feedback from @ALRubinger and @saeedjabbar.
* Added two **NEW** Open Questions.
* Implemented the end-to-end community contributed docs workflow.
* Implemented Live Preview for local doc editing.

## End-to-end Worklow Notes

Two repos are used to mock the TBD Developer Site and the Community Docs repos:
- TBD Developer Site: https://github.com/frankhinek/prototype-tbdocusaurus
- TBD Community Docs: https://github.com/frankhinek/prototype-tbdocs

The steps for authoring new content are:

1. A contributor would start by forking the [TBD Community Docs](https://github.com/frankhinek/prototype-tbdocs) repo, cloning locally, and adding/editing content.
2. There's now a Live Preview utility so contributors can see how their content will look once its published: [tbdoc Live Preview](https://github.com/frankhinek/prototype-tbdocs).
3. Once satisfied with their additions/changes, a contributor commits their changes to their fork and submits a PR to the [TBD Community Docs](https://github.com/frankhinek/prototype-tbdocs) repo.
4. TBD and community members review, and if needed, provide feedback or request changes.
5. Once the PR is approved and merged to the [TBD Community Docs](https://github.com/frankhinek/prototype-tbdocs) main branch, GitHub Actions automatically kicks off an [Open Dev Site PR workflow](https://github.com/frankhinek/prototype-tbdocs/blob/main/.github/workflows/open-dev-site-pr.yml) to process the Markdown content and open a PR in the [TBD Developer Site](https://github.com/frankhinek/prototype-tbdocusaurus) mockup.
6. The TBD Developer Site PR contains the committed files, a link to the original approved merge, and also a Netlify Deploy Preview for the team managing the TBD Developer Site to review the visual appearance of the contributions.
7. If the changes are accepted, the PR is approved/merged and then the typical publishing process for the TBD Dev Site would kick off.

Here's an example:

- Commit to community docs mock repo: https://github.com/frankhinek/prototype-tbdocs/pull/13/commits/669877e53a3416fe0f81e23fa8742551dc31708a
- GitHub Actions ran: https://github.com/frankhinek/prototype-tbdocs/actions/runs/3027820249
- PR opened on Dev Site mock repo: https://github.com/frankhinek/prototype-tbdocusaurus/pull/5

**Comment by ALRubinger** [comment]
This is amazing. Thanks for the patience while I've found the time to grok the comprehensive docs and instructions you've laid out here. All ran great here and I totally see where you're going with this workflow. Eager to get it into the repo and hook things together, and I've granted you admin access here.

I've written up some more details on the forum thread. 

To your open questions:

>Docusaurus is used for [developer.tbd.website](https://developer.tbd.website/), which supports [MDX](https://mdxjs.com/docs/what-is-mdx/). Should we support including JSX in markdown content for community-contributed docs?
> This one may get murky as I believe it's tied into how Docusaurus handles MDX and pulling in external Markdown. Long term we may want to offer docs authors greater flexibility in styling by exposing JSX directly. In service of MVP I'd be thrilled to start by releasing a system which lets us author in Markdown, embed code and images as you've shown, and get things published on the Developer Site. WDYT?

Exactly as you suggest. Let's get some Markdown where we can embed code and images and iterate from there. That alone is a huge win for bootstrapping contribution and getting it published.

> When new documentation is approved and PRs merged, GitHub Actions can be triggered to build the doc artifacts and potentially trigger publishing to [developer.tbd.website](https://develop.tbd.website/). Should this workflow run in the main Docusaurus site repo or the community-contributed docs repo? What additional steps can we take to automate the publication process and minimize human intervention?
> Hot takes: chained builds? The community-contributed docs repo can run builds to compose the output Markdown, and then we trigger a PR into the Developer Site repo to update/add the newly-generated page? This would let us always revert the Developer Site if we needed, and makes clear when a new piece of content is to be published, while letting the docs repo handle its own build to generate the markdown.

Love the chained builds approach and it's given me success before. Because: then we've cleanly decoupled the docs contribution piece into its own workflow, and the publishing step will show us in each commit exactly what's to be pushed live on the dev site. Also makes for a super clean audit trail in the case we need to roll back changes on the Dev Site. I'm hoping this is a really easy contribution path for people by letting them focus only on the writing and docs generation, knowing the grander details of publishing will just (waves magic wand) be handled for them through automation.

@saeedjabbar had an open question too:

> Q: Does the current developer.tbd.website Docusaurus instance already have i18n configured, and if so, can you share a few details that can guide how to structure community contributed docs?

Docusaurus has i18n support but we're not taking advantage of it yet, and I'm thinking that once we get this plugged in, i18n is an awesome thing to enable and let the community run with for translations. I was once astounded at how eager folks were to translate the [Arquillian Quides](https://arquillian.org/guides/) into so many languages.
